File: packages/cli/src/telemetry/emitter.ts
import type { BuiltTelemetryEvent } from './events.js';

export interface TelemetryEmitterEnv {
  KTX_TELEMETRY_DEBUG?: string;
  KTX_TELEMETRY_ENDPOINT?: string;
}

export interface TelemetrySink {
  write(chunk: string): void;
}

type PostHogClient = {
  capture(event: {
    distinctId: string;
    event: string;
    properties: Record<string, unknown>;
    groups?: Record<string, string>;
  }): void;
  shutdown(): Promise<void> | void;
};

// PostHog public project ingestion key — safe to embed; capture-only, no read access.
const POSTHOG_PROJECT_API_KEY = 'phc_xbvZpbu8ZNLnogTbY7MEMWhCF2rzzApYsDndjKaRBXXx'; // pragma: allowlist secret
const POSTHOG_HOST = 'https://us.i.posthog.com';
const SHUTDOWN_TIMEOUT_MS = 1500;

let clientPromise: Promise<PostHogClient | null> | undefined;

function telemetryHost(env: TelemetryEmitterEnv, explicitHost?: string): string {
  return explicitHost ?? env.KTX_TELEMETRY_ENDPOINT ?? POSTHOG_HOST;
}

function telemetryProjectApiKey(explicitProjectApiKey?: string): string {
  return explicitProjectApiKey ?? POSTHOG_PROJECT_API_KEY;
}

function liveTelemetryConfigured(projectApiKey: string, host: string): boolean {
  return projectApiKey.trim() !== '' && host.trim() !== '';
}

async function getPostHogClient(projectApiKey: string, host: string): Promise<PostHogClient | null> {
  if (!liveTelemetryConfigured(projectApiKey, host)) {
    return null;
  }

  clientPromise ??= import('posthog-node')
    .then(({ PostHog }) => new PostHog(projectApiKey, { host, flushAt: 1, flushInterval: 0 }))
    .catch(() => null);

  return await clientPromise;
}

function debugEnabled(env: TelemetryEmitterEnv): boolean {
  return env.KTX_TELEMETRY_DEBUG === '1';
}

function writeDebugPayload(input: {
  event: BuiltTelemetryEvent;
  distinctId: string;
  projectId?: string;
  stderr: TelemetrySink;
}): void {
  input.stderr.write(
    `[telemetry] ${JSON.stringify({
      distinctId: input.distinctId,
      event: input.event.name,
      properties: input.event.properties,
      groups: input.projectId ? { project: input.projectId } : undefined,
    })}\n`,
  );
}

export async function trackTelemetryEvent(input: {
  event: BuiltTelemetryEvent;
  distinctId: string;
  projectId?: string;
  env?: TelemetryEmitterEnv;
  stderr: TelemetrySink;
  projectApiKey?: string;
  host?: string;
}): Promise<void> {
  const env = input.env ?? process.env;

  if (debugEnabled(env)) {
    writeDebugPayload(input);
    return;
  }

  const projectApiKey = telemetryProjectApiKey(input.projectApiKey);
  const host = telemetryHost(env, input.host);
  const client = await getPostHogClient(projectApiKey, host);
  if (!client) {
    return;
  }

  try {
    client.capture({
      distinctId: input.distinctId,
      event: input.event.name,
      properties: input.event.properties,
      groups: input.projectId ? { project: input.projectId } : undefined,
    });
  } catch {
    return;
  }
}

export async function shutdownTelemetryEmitter(): Promise<void> {
  const client = await clientPromise;
  if (!client) {
    return;
  }

  await Promise.race([
    Promise.resolve(client.shutdown()).catch(() => undefined),
    new Promise<void>((resolve) => {
      setTimeout(resolve, SHUTDOWN_TIMEOUT_MS);
    }),
  ]);
}

/** @internal */
export function __resetTelemetryEmitterForTests(): void {
  clientPromise = undefined;
}

File: python/ktx-daemon/src/ktx_daemon/telemetry/emitter.py
from __future__ import annotations

import json
import os
import sys
from pathlib import Path
from typing import Any
from collections.abc import Mapping

from ktx_daemon.telemetry.events import build_telemetry_event
from ktx_daemon.telemetry.identity import load_telemetry_identity

# PostHog public project ingestion key - safe to embed; capture-only, no read access.
POSTHOG_PROJECT_API_KEY = (
    "phc_xbvZpbu8ZNLnogTbY7MEMWhCF2rzzApYsDndjKaRBXXx"  # pragma: allowlist secret
)
POSTHOG_HOST = "https://us.i.posthog.com"


def _host(env: Mapping[str, str]) -> str:
    return env.get("KTX_TELEMETRY_ENDPOINT") or POSTHOG_HOST


def _live_configured(host: str) -> bool:
    return bool(POSTHOG_PROJECT_API_KEY.strip() and host.strip())


def _debug_enabled(env: Mapping[str, str]) -> bool:
    return env.get("KTX_TELEMETRY_DEBUG") == "1"


def _scrub_error_class(error: BaseException) -> str | None:
    name = type(error).__name__
    if len(name) > 80:
        return None
    if any(marker in name for marker in ("/", "\\", "@", "://")):
        return None
    if not name[:1].isupper() or not name.replace("_", "").isalnum():
        return None
    return name


def error_class(error: BaseException) -> str | None:
    return _scrub_error_class(error)


def track_telemetry_event(
    name: str,
    fields: dict[str, Any],
    *,
    project_id: str | None = None,
    home_dir: Path | None = None,
    env: Mapping[str, str] | None = None,
) -> None:
    source_env = env or os.environ
    identity = load_telemetry_identity(home_dir=home_dir, env=source_env)
    if not identity.enabled or not identity.install_id:
        return

    try:
        event = build_telemetry_event(name, fields)
    except ValueError:
        return

    groups = {"project": project_id} if project_id else None

    if _debug_enabled(source_env):
        sys.stderr.write(
            "[telemetry] "
            + json.dumps(
                {
                    "distinctId": identity.install_id,
                    "event": event["event"],
                    "properties": event["properties"],
                    "groups": groups,
                },
                sort_keys=True,
            )
            + "\n"
        )
        return

    host = _host(source_env)
    if not _live_configured(host):
        return

    try:
        from posthog import Posthog

        client = Posthog(
            POSTHOG_PROJECT_API_KEY,
            host=host,
            flush_at=1,
            flush_interval=0,
            sync_mode=True,
            timeout=1,
        )
        client.capture(
            event=event["event"],
            distinct_id=identity.install_id,
            properties=event["properties"],
            groups=groups,
        )
        client.shutdown()
    except Exception:
        return
